Framework/Core/include/Framework/AnalysisTask.h

```diff
@@ -392,146 +393,236 @@ struct AnalysisDataProcessorBuilder {
     return std::tuple<>{};
   }
 
+  template <typename G, typename... A>
+  struct GroupSlicer {
+    using grouping_t = std::decay_t<G>;
+    GroupSlicer(G& gt, std::tuple<A...>& at)
+      : max{gt.size()},
+        mBegin{GroupSlicerIterator(gt, at)}
+    {
+    }
+
+    struct GroupSlicerSentinel {
+      int64_t position;
+    };
+
+    struct GroupSlicerIterator {
+      using associated_pack_t = framework::pack<A...>;
+
+      GroupSlicerIterator() = default;
+      GroupSlicerIterator(GroupSlicerIterator const&) = default;
+      GroupSlicerIterator(GroupSlicerIterator&&) = default;
+      GroupSlicerIterator& operator=(GroupSlicerIterator const&) = default;
+      GroupSlicerIterator& operator=(GroupSlicerIterator&&) = default;
+
+      GroupSlicerIterator(G& gt, std::tuple<A...>& at)
+        : mAt{&at},
+          groupingElement{gt.begin()},
+          position{0}
+      {
+        using groupingMetadata = typename aod::MetadataTrait<G>::metadata;
+        auto indexColumnName = std::string("f") + groupingMetadata::label() + "ID";
+        arrow::compute::FunctionContext ctx;
+        /// prepare slices and offsets for all associated tables that have index
+        /// to grouping table
+        ///
+        auto splitter = [&](auto&& x) {
+          if (hasIndexTo<std::decay_t<G>>(typename std::decay_t<decltype(x)>::persistent_columns_t{})) {
+            auto result = o2::framework::sliceByColumn(&ctx, indexColumnName,
+                                                       x.asArrowTable(),
+                                                       &groups[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})],
+                                                       &offsets[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})]);
+            if (result.ok() == false) {
+              throw std::runtime_error("Cannot split collection");
+            }
+          }
+        };
+
+        std::apply(
+          [&](auto&&... x) -> void {
+            (splitter(x), ...);
+          },
+          at);
+        /// extract selections from filtered associated tables
+        auto extractor = [&](auto&& x) {
+          if constexpr (soa::is_soa_filtered_t<std::decay_t<decltype(x)>>::value) {
+            selections[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})] = &x.getSelectedRows();
+            starts[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})] = selections[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})]->begin();
+            offsets[framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})].push_back(std::get<std::decay_t<decltype(x)>>(at).tableSize());
+          }
```

Comment:
How about a `using someType = std::decay_t<decltype(x)>` in the context of the lambda? Also isn't
`framework::has_type_at<std::decay_t<decltype(x)>>(associated_pack_t{})` repeated three times?

Reply:
Fixed.